### PR TITLE
i#5539: Split L0_filter into two separate options

### PR DESCRIPTION
**Pull request by abhinav92003**
Splits the -L0_filter option into two separate ones, -L0I_filter and -L0D_filter,
for instruction and data address filtering respectively. This is to support
use-cases where we want to filter only one but not the other.

The existing -L0_filter option is still supported, for backward compatibility.
We simply sets the -L0I_filter and -L0D_filter flags when we see this for code
simplicity.

We disable optimizations when either -L0D_filter or -L0I_filter is set.

Introduces new file types OFFLINE_FILE_TYPE_IFILTERED and
OFFLINE_FILE_TYPE_DFILTERED to denote raw files where instructions
or data addresses respectively are filtered. We need
OFFLINE_FILE_TYPE_IFILTERED so that we can know whether bb
instrs are separate, which happens only with L0I_filter.

We could have done without OFFLINE_FILE_TYPE_DFILTERED and set
the existing OFFLINE_FILE_TYPE_FILTERED when either instrs or data
addresses were filtered. But we want to continue supporting the older trace
versions where OFFLINE_FILE_TYPE_FILTERED also means that bb
instrs are separate due to instr filtering. For symmetry, we just add the new
OFFLINE_FILE_TYPE_DFILTERED type.

Also adds a new sub-type for OFFLINE_TYPE_EXTENDED to denote the new
header format which has the file type and file version fields swapped. This is
required since we are out of bits to store the offline file type in the header. The
new enum value is named same as the old one (OFFLINE_EXT_TYPE_HEADER),
while the old one is now renamed to (OFFLINE_EXT_TYPE_HEADER_DEPRECATED).

Also bumps up the version for offline files to
OFFLINE_FILE_VERSION_HEADER_FIELDS_SWAP, since we have the
new header entry now.

Adds tests for some combinations of these options. Changes the
-simulator_type to basic_counts as we really want to look at counts in these
tests.

Issue: #5539

### CODE REVIEW

**Comment by abhinav92003** [comment]
run arm tests

File: clients/drcachesim/tracer/tracer.cpp [hunk 1]
```diff
@@ -3216,17 +3235,19 @@ drmemtrace_client_main(client_id_t id, int argc, const char *argv[])
...
         DR_ASSERT(MAX_INSTRU_SIZE >= sizeof(offline_instru_t));
         placement = dr_global_alloc(MAX_INSTRU_SIZE);
-        instru = new (placement) offline_instru_t(
-            insert_load_buf_ptr, op_L0_filter.get_value(), &scratch_reserve_vec,
-            file_ops_func.write_file, module_file, op_disable_optimizations.get_value());
+        instru = new (placement)
+            offline_instru_t(insert_load_buf_ptr, op_L0I_filter.get_value(),
+                             &scratch_reserve_vec, file_ops_func.write_file, module_file,
+                             op_disable_optimizations.get_value() ||
+                                 op_L0D_filter.get_value() || op_L0I_filter.get_value());
```
Comment (derekbruening) [inline clients/drcachesim/tracer/tracer.cpp:3242]: -use_physical sets disable_optimizations to true: should filtering do the same?  Then the file type would have the no-opt flag set (which might simplify raw2trace logic?).

**Review by derekbruening — APPROVED** [review]

File: api/docs/release.dox [hunk 2]
```diff
@@ -149,6 +149,14 @@ Further non-compatibility-affecting changes include:
    dr_hashtable_destroy(), dr_hashtable_clear(), dr_hashtable_lookup(),
    dr_hashtable_add(), dr_hashtable_remove().
  - Added a new #TRACE_MARKER_TYPE_PAGE_SIZE record to drcachesim offline traces.
+ - Added new options -L0I_filter and -L0D_filter that allow enabling online
+   filtering for only instr or only data entries respectively. The old option
+   -L0_filter is deprecated but still supported for backward compatibility. It
+   simply sets both the new options. Related to this, also added a new sub-type
+   for OFFLINE_TYPE_EXTENDED to denote the new header format which has the
```
[the remark this answers is not in the document]
Comment (abhinav92003) [reply api/docs/release.dox]: Removed.

File: api/docs/release.dox [hunk 3]
```diff
@@ -149,6 +149,14 @@ Further non-compatibility-affecting changes include:
    dr_hashtable_destroy(), dr_hashtable_clear(), dr_hashtable_lookup(),
    dr_hashtable_add(), dr_hashtable_remove().
  - Added a new #TRACE_MARKER_TYPE_PAGE_SIZE record to drcachesim offline traces.
+ - Added new options -L0I_filter and -L0D_filter that allow enabling online
```
[the remark this answers is not in the document]
Comment (abhinav92003) [reply api/docs/release.dox]: Done.

File: clients/drcachesim/common/trace_entry.h [hunk 4]
```diff
@@ -510,7 +517,9 @@ typedef enum {
  */
 typedef enum {
     OFFLINE_FILE_TYPE_DEFAULT = 0x00,
-    OFFLINE_FILE_TYPE_FILTERED = 0x01, /**< Addresses filtered online. */
+    // This is deprecated. Newer trace files use OFFLINE_FILE_TYPE_IFILTERED
```
[the remark this answers is not in the document]
Comment (abhinav92003) [reply clients/drcachesim/common/trace_entry.h]: This is removed now.

File: clients/drcachesim/common/trace_entry.h [hunk 5]
```diff
@@ -519,11 +528,9 @@ typedef enum {
     OFFLINE_FILE_TYPE_ARCH_X86_64 = 0x40,      /**< Recorded on x86 (64-bit). */
     OFFLINE_FILE_TYPE_ARCH_ALL = OFFLINE_FILE_TYPE_ARCH_AARCH64 |
         OFFLINE_FILE_TYPE_ARCH_ARM32 | OFFLINE_FILE_TYPE_ARCH_X86_32 |
-        OFFLINE_FILE_TYPE_ARCH_X86_64, /**< All possible architecture types. */
-    // For raw files, this is currently stored in an 8-bit field.
-    // If we run out of flags we should swap the version to be in valueB and
-    // the flags in valueA, leaving the bottom few bits of valueA for compatibility
-    // with old versions.
+        OFFLINE_FILE_TYPE_ARCH_X86_64,   /**< All possible architecture types. */
+    OFFLINE_FILE_TYPE_IFILTERED = 0x80,  /**< Instr addresses filtered online. */
```
[the remark this answers is not in the document]
Comment (abhinav92003) [reply clients/drcachesim/common/trace_entry.h]: Done

File: clients/drcachesim/common/trace_entry.h [hunk 6]
```diff
@@ -510,7 +517,9 @@ typedef enum {
  */
 typedef enum {
     OFFLINE_FILE_TYPE_DEFAULT = 0x00,
-    OFFLINE_FILE_TYPE_FILTERED = 0x01, /**< Addresses filtered online. */
+    // This is deprecated. Newer trace files use OFFLINE_FILE_TYPE_IFILTERED
+    // and OFFLINE_FILE_TYPE_DFILTERED.
+    OFFLINE_FILE_TYPE_FILTERED_DEPRECATED = 0x01, /**< Addresses filtered online. */
```
[the remark this answers is not in the document]
Comment (abhinav92003) [reply clients/drcachesim/common/trace_entry.h]: Reverted.

File: clients/drcachesim/tracer/tracer.cpp (hunk 1 above)
Comment (abhinav92003) [reply clients/drcachesim/tracer/tracer.cpp]: Done.

File: clients/drcachesim/tests/filter-l0.templatex [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+     .* total icache flushes
+     .* total dcache flushes
+           1 total threads
+     .* total scheduling markers
+     .* total transfer markers
+     .* total function id markers
+     .* total function return address markers
+     .* total function argument markers
+     .* total function return value markers
+     .* total other markers
```
[the remark this answers is not in the document]
Comment (abhinav92003) [reply clients/drcachesim/tests/filter-l0.templatex]: Added generic template files for basic count tests.

[the remark this answers is not in the document]
Comment (abhinav92003) [reply clients/drcachesim/tests/filter-l0.templatex]: I did a quick scan to see if any existing templates also do the same, but didn't find any.

File: clients/drcachesim/tests/filter-no-d.templatex [hunk 8]
```diff
@@ -1,29 +1,39 @@
...
+     .* total icache flushes
+     .* total dcache flushes
+           1 total threads
+     .* total scheduling markers
+     .* total transfer markers
+     .* total function id markers
+     .* total function return address markers
+     .* total function argument markers
+     .* total function return value markers
+     .* total other markers
```
[the remark this answers is not in the document]
Comment (abhinav92003) [reply clients/drcachesim/tests/filter-no-d.templatex:22]: Done.

File: clients/drcachesim/tracer/tracer.cpp (hunk 1 above)
Comment (abhinav92003) [reply clients/drcachesim/tracer/tracer.cpp]: I was able to simply the raw2trace logic just a bit. In `analyze_elidable_addresses`, now I don't need to check `OFFLINE_FILE_TYPE_IFILTERED` and `OFFLINE_FILE_TYPE_DFILTERED`. I still check `OFFLINE_FILE_TYPE_FILTERED` as that's a legacy type. 

I still need to check the new offline file types for instrs_is_separate and the assert statement for `memref.use_remembered_base`.

File: clients/drcachesim/common/trace_entry.h [hunk 9]
```diff
@@ -500,7 +506,8 @@ typedef enum {
 #define OFFLINE_FILE_VERSION_OLDEST_SUPPORTED OFFLINE_FILE_VERSION_NO_ELISION
 #define OFFLINE_FILE_VERSION_ELIDE_UNMOD_BASE 3
 #define OFFLINE_FILE_VERSION_KERNEL_INT_PC 4
-#define OFFLINE_FILE_VERSION OFFLINE_FILE_VERSION_KERNEL_INT_PC
+#define OFFLINE_FILE_VERSION_HEADER_FIELDS_SWAP 5
```
[the remark this answers is not in the document]
Comment (abhinav92003) [reply clients/drcachesim/common/trace_entry.h:509]: Done.

File: clients/drcachesim/common/trace_entry.h [hunk 10]
```diff
@@ -517,9 +517,11 @@ typedef enum {
  */
 typedef enum {
     OFFLINE_FILE_TYPE_DEFAULT = 0x00,
-    // This is deprecated. Newer trace files use OFFLINE_FILE_TYPE_IFILTERED
-    // and OFFLINE_FILE_TYPE_DFILTERED.
-    OFFLINE_FILE_TYPE_FILTERED_DEPRECATED = 0x01, /**< Addresses filtered online. */
+    /**
+     * DEPRECATED: Addresses filtered online. Newer trace files use
+     * OFFLINE_FILE_TYPE_IFILTERED and OFFLINE_FILE_TYPE_DFILTERED.
```
Comment (derekbruening) [inline clients/drcachesim/common/trace_entry.h:522]: Use # for links for both.

Comment (abhinav92003) [reply clients/drcachesim/common/trace_entry.h]: Done.

**Pull request merged**